Orange/widgets/data/owpythonscript.py

```diff
@@ -645,7 +667,7 @@ def saveScript(self):
             f.close()
 
     def initial_locals_state(self):
-        d = dict([(i.name, getattr(self, i.name, None)) for i in self.inputs])
+        d = dict([(name, getattr(self, name, None)) for name in self.inputs_names])
```

Comment:
No, no, `in_data` has to be in a dictionary, but doesn't itself have to be a dictionary. `locals` is a dict with local variables (name bindings, actually), that is, keys are variable names and values are their values. That is, `d` has to be a dict, but its values can be anything.

Here:

change the `inputs_names` back to just

        inputs_names = ("in_data", "in_learner", "in_classifier", "in_object")

and remove singulars in `__init__`:

        self.in_data = {}
        self.in_learner = {}
        self.in_classifier = {}
        self.in_object = {}

Change signal handlers to

    def setExampleTable(self, et, id=None):
        self._set_dict(self.in_data, et, id)

    def setDistanceMatrix(self, dm):
        self.in_distance = dm

    def setLearner(self, learner, id=None):
        self._set_dict(self.in_learner, learner, id)

    def setClassifier(self, classifier, id=None):
        self._set_dict(self.in_classifier, classifier, id)

    def setObject(self, obj, id=None):
        self._set_dict(self.in_object, obj, id)

and construct locals with

    def initial_locals_state(self):
        d = dict.fromkeys(o.name for o in self.outputs)
        for name in self.inputs_names:
            value = getattr(self, name, None)
            all_values = list(value.values()) if value is not None else []
            one_value = all_values[0] if len(all_values) == 1 else None
            d[name + 's'] = all_values
            d[name] = one_value
        return d

If I haven't overlooked something, this gives the behaviour I wanted. Please check, and also see whether `id` can be a dict (my other comment).